src/device/router_device_worker.erl

```diff
@@ -1278,12 +1278,27 @@ validate_join(
         true ->
             {error, bad_nonce};
         false ->
-            PayloadSize = erlang:byte_size(Payload),
-            PHash = blockchain_helium_packet_v1:packet_hash(Packet),
-            case maybe_charge(Device, PayloadSize, PubKeyBin, PHash, OfferCache) of
-                {error, _} = Error ->
-                    Error;
-                {ok, Balance, Nonce} ->
+            case router_utils:get_env_bool(charge_joins, true) of
+                true ->
+                    PayloadSize = erlang:byte_size(Payload),
+                    PHash = blockchain_helium_packet_v1:packet_hash(Packet),
+                    case maybe_charge(Device, PayloadSize, PubKeyBin, PHash, OfferCache) of
+                        {error, _} = Error ->
+                            Error;
+                        {ok, Balance, Nonce} ->
+                            {ok, UpdatedDevice, JoinAcceptArgs} = handle_join(
+                                Packet,
+                                PubKeyBin,
+                                Region,
+                                APIDevice,
+                                AppKey,
+                                Device
+                            ),
+                            {ok, UpdatedDevice, DevNonce, JoinAcceptArgs, {Balance, Nonce}}
+                    end;
+                false ->
+                    OrgID = maps:get(organization_id, router_device:metadata(Device), undefined),
+                    {Balance, Nonce} = router_console_dc_tracker:current_balance(OrgID),
```

Comment:
We need to heck that returning current balance won't affect Console

Reply:
https://github.com/helium/console/blob/3763e1b38f69b099955526a28cb80762e5113e96/lib/console/etl_error_worker.ex#L96C19-L114

balance is only updated if `used` has a value `> 0`

Reply:
If anything, it will be a DB call that sets the balance to what it already is.
There's a change in https://github.com/helium/console/pull/1279 to only update the balance if DC was used.